docs/provisioning.rst

```diff
@@ -0,0 +1,355 @@
+              "max_cpus": 4
+          }
+        }
+      },
+
+Kernel provisioner authors implement their provisioners by deriving from
+:class:`KernelProvisionerBase` and expose their provisioner for consumption
+via entry-points:
+
+.. code:: python
+
+    'jupyter_client.kernel_provisioners': [
+                'k8s-provisioner = my_package:K8sProvisioner',
+            ],
+
+Backwards Compatibility
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Prior to this release, no ``kernel.json`` (kernelspec) will contain a
+provisioner entry, yet the framework is now based on using provisioners.
+As a result, when a ``kernel_provisioner`` stanza is **not** present in
+a selected kernelspec, jupyter client will, by default, use the built-in
+``LocalProvisioner`` implementation as its provisioner. This provisioner
+retains today's local kernel functionality. It can also be subclassed
+for those provisioner authors wanting to extend the functionality of
+local kernels. The result of launching a kernel in this manner is
+equivalent to the following stanza existing in the ``kernel.json`` file:
+
+.. code:: JSON
+
+      "metadata": {
+        "kernel_provisioner": {
+          "provisioner_name": "local-provisioner",
+          "config": {
+          }
+        }
+      },
+
+Should a given installation wish to use a *different* provisioner as
+their "default provisioner" (including subclasses of
+``LocalProvisioner``), they can do so by specifying a value for
+``KernelProvisionerFactory.default_provisioner_name``.
+
+Implementing a custom provisioner
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The impact of Kernel Provisioning is that it enables the ability to
+implement custom kernel provisioners to manage a kernel's lifecycle
+within any runtime environment. There are currently two approaches by
+which that can be accomplished, extending the ``KernelProvisionerBase``
+class or extending the built-in class - ``LocalProvisioner``. As more
+provisioners are introduced, some may be implemented in an abstract
+sense, from which specific implementations can be authored.
+
+Extending ``LocalProvisioner``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you're interested in running kernels locally and yet adjust their
+behavior, there's a good chance you can simply extend
+``LocalProvisioner`` via subclassing. This amounts to deriving from
+``LocalProvisioner`` and overriding appropriate methods to provide your
+custom functionality.
+
+In this example, RBACProvisioner will verify whether the current user is
+in the role meant for this kernel. If not, an exception will be thrown.
+
+.. code:: python
+
+    class RBACProvisioner(LocalProvisioner):
+
+        role: str = Unicode(config=True)
+
+        async def pre_launch(self, **kwargs: Any) -> Dict[str, Any]:
+
+            if not self.user_in_role(self.role):
+                raise PermissionError(f"User is not in role {self.role} and "
+                                      f"cannot launch this kernel.")
+
```

Comment:
Maybe clearly state that the `user_in_role` method has to be implemented by the Provisioner and is not part of the Provisioner API.

Reply:
That makes sense - update pushed.